src/lib/shell/Engine.cpp

```diff
@@ -56,47 +59,90 @@ int Engine::Init()
 
 void Engine::ForEachCommand(shell_command_iterator_t * on_command, void * arg)
 {
-    for (unsigned i = 0; i < _commandSetCount; i++)
+    for (auto * _command : _commandSet)
     {
-        for (unsigned j = 0; j < _commandSetSize[i]; j++)
+        if (on_command(_command, arg) != CHIP_NO_ERROR)
         {
-            if (on_command(&_commandSet[i][j], arg) != CHIP_NO_ERROR)
-            {
-                return;
-            }
+            return;
         }
     }
 }
 
-void Engine::RegisterCommands(shell_command_t * command_set, unsigned count)
+void Engine::RegisterCommands(shell_command_t * command_set, unsigned count, const char * prefix)
 {
-    if (_commandSetCount >= CHIP_SHELL_MAX_MODULES)
+    if (this == &Engine::Root() || prefix == nullptr)
     {
-        ChipLogError(Shell, "Max number of modules reached\n");
+        prefix = "";
+    }
+
+    if (_commandSet.size() + count > CHIP_SHELL_MAX_MODULES)
```

Comment:
Thanks a lot for the review and guidance. I updated the implementation to avoid STL. 

I initially also thought about the code size but it appeared there wasn't obvious change w/wo STL to me.